src/mstr.c

```diff
@@ -0,0 +1,518 @@
+/*
+ * Copyright Redis Ltd. 2024 - present
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2)
+ * or the Server Side Public License v1 (SSPLv1).
+ */
+
+#include <string.h>
+#include <assert.h>
+#include "sdsalloc.h"
+#include "mstr.h"
+#include "stdio.h"
+
+#define NULL_SIZE 1
+
+static inline char mstrReqType(size_t string_size);
+static inline int mstrHdrSize(char type);
+static inline int mstrSumMetaLen(mstrKind *k, mstrFlags flags);
+static inline size_t mstrAllocLen(const mstr s, struct mstrKind *kind);
+
+/*** mstr API ***/
+
+/* Create mstr without any metadata attached, based on string 'initStr'.
+ * - If initStr equals NULL, then only allocation will be made.
+ * - string of mstr is always null-terminated.
+ */
+mstr mstrNew(const char *initStr, size_t lenStr) {
```

Comment:
Maybe we can make it call mstrNewWithMeta(), the code is mostly the same.

Reply:
`mstrNew(str, len)` and `mstrNewWithMeta(kind, str, len, metaFlags)` are not  the same. The first allocates just bare string. The second allocates a string with flags, yet without any metadata structures allocated.